### PR TITLE
refactoring classes AsyncBatchDestroyJob and AsyncBatchUpdateJob

### PR DESCRIPTION
**Pull request by Ivanov-Anton**
refactoring classes according to the interface ActiveJob

### CODE REVIEW

File: app/jobs/async_batch_update_job.rb [hunk 1]
```diff
@@ -1,19 +1,16 @@
 # frozen_string_literal: true
 
-class AsyncBatchUpdateJob
+class AsyncBatchUpdateJob < ApplicationJob
```
Comment (gigorok) [inline app/jobs/async_batch_update_job.rb:3]: call it in a new way

File: app/jobs/async_batch_update_job.rb [hunk 2]
```diff
@@ -1,19 +1,16 @@
...
   BATCH_SIZE = 1000
 
   attr_reader :model_class, :sql_query, :changes, :who_is
 
-  def initialize(model_class, sql_query, changes, who_is)
+  def perform(model_class, sql_query, changes, who_is)
     @model_class = model_class
     @sql_query = sql_query
     @changes = changes
     @who_is = who_is
```
Comment (gigorok) [inline app/jobs/async_batch_update_job.rb:13]: looks like only sql_query reader is needed

File: app/jobs/async_batch_update_job.rb (hunk 1 above)
Comment (gigorok) [reply app/jobs/async_batch_update_job.rb:3]: AsyncBatchUpdateJob.perform_later(...)

File: app/jobs/async_batch_update_job.rb [hunk 3]
```diff
@@ -1,19 +1,16 @@
...
-class AsyncBatchUpdateJob
+class AsyncBatchUpdateJob < ApplicationJob
   include BatchJobsLog
   BATCH_SIZE = 1000
 
   attr_reader :model_class, :sql_query, :changes, :who_is
 
-  def initialize(model_class, sql_query, changes, who_is)
+  def perform(model_class, sql_query, changes, who_is)
     @model_class = model_class
```
Comment (gigorok) [inline app/jobs/async_batch_update_job.rb:10]: @model_class = model_class.constantize

File: app/jobs/async_batch_update_job.rb [hunk 4]
```diff
@@ -1,19 +1,16 @@
...
     @sql_query = sql_query
     @changes = changes
     @who_is = who_is
-  end
-
-  def perform
     set_audit_log_data
     model_class.constantize.transaction do
```
Comment (gigorok) [inline app/jobs/async_batch_update_job.rb:15]: remove constantize

File: lib/resource_dsl/acts_as_async_destroy.rb [hunk 5]
```diff
@@ -10,7 +10,7 @@ def acts_as_async_destroy(model_class)
       scoped_collection_action :async_destroy,
                                title: 'Delete batch',
                                if: proc { authorized?(:batch_destroy, resource_klass) } do
-        Delayed::Job.enqueue AsyncBatchDestroyJob.new(model_class,
+        Delayed::Job.enqueue AsyncBatchDestroyJob.perform_later(model_class,
```
Comment (gigorok) [inline lib/resource_dsl/acts_as_async_destroy.rb:13]:
```suggestion
        AsyncBatchDestroyJob.perform_later(model_class,
```

File: lib/resource_dsl/acts_as_async_update.rb [hunk 6]
```diff
@@ -11,7 +11,7 @@ def acts_as_async_update(model_class, attrs_to_update)
                                class: 'scoped_collection_action_button ui',
                                form: attrs_to_update,
                                if: proc { authorized?(:batch_destroy, resource_klass) } do
-        Delayed::Job.enqueue AsyncBatchUpdateJob.new(model_class,
+        Delayed::Job.enqueue AsyncBatchUpdateJob.perform_later(model_class,
```
Comment (gigorok) [inline lib/resource_dsl/acts_as_async_update.rb:14]:
```suggestion
        AsyncBatchUpdateJob.perform_later(model_class,
```

File: lib/resource_dsl/acts_as_async_destroy.rb [hunk 7]
```diff
@@ -10,10 +10,9 @@ def acts_as_async_destroy(model_class)
       scoped_collection_action :async_destroy,
                                title: 'Delete batch',
                                if: proc { authorized?(:batch_destroy, resource_klass) } do
-        Delayed::Job.enqueue AsyncBatchDestroyJob.new(model_class,
+        AsyncBatchDestroyJob.set(queue: 'batch_actions').perform_later(model_class,
```
Comment (gigorok) [inline lib/resource_dsl/acts_as_async_destroy.rb:13]: remove set(queue: 'batch_actions'). already has such queue

File: app/jobs/async_batch_destroy_job.rb [hunk 8]
```diff
@@ -29,4 +26,11 @@ def reschedule_at(_time, _attempts)
   def max_attempts
     3
   end
+
+  around_perform do |_job, block|
+    block.call
+    success
+  rescue Exception => error
```
Comment (gigorok) [inline app/jobs/async_batch_destroy_job.rb:33]: rescue StandardError => error

File: lib/batch_jobs_log.rb [hunk 9]
```diff
@@ -1,20 +1,21 @@
...
-  def failure(job)
+  def failure(error)
     LogicLog.create!(
-      source: "#{self.class} #{job.id}",
+      source: "#{self.class} #{job_id}",
       level: 0,
-      msg: job.last_error
+      msg: "ERROR: #{error.message}\nchanges: #{arguments[2]}\nclass: #{model_class}\nqueue: #{queue_name}\nsql: #{sql_query}"
     )
+    raise error.class
```
Comment (gigorok) [inline lib/batch_jobs_log.rb:18]: raise error to save original error message

File: lib/batch_jobs_log.rb [hunk 10]
```diff
@@ -1,20 +1,39 @@
...
-  def failure(job)
+  def failure(error)
     LogicLog.create!(
-      source: "#{self.class} #{job.id}",
+      source: "#{self.class} #{delayed_job_record_id}",
       level: 0,
-      msg: job.last_error
+      msg: "Error: #{error.message}\nchanges: #{arguments[2]}\nclass: #{arguments[0]}\nqueue: #{queue_name}\nsql: #{arguments[1]}"
     )
+    raise error
```
Comment (senid231) [inline lib/batch_jobs_log.rb:36]:
```
# frozen_string_literal: true

module BatchJobsLog
  extend ActiveSupport::Concern

  included do
    around_perform :with_logic_log
  end

  private

  def with_logic_log
    Rails.logger.info { "Delayed::Worker.plugins #{Delayed::Worker.plugins}" }
    result = yield
    success
    result
  rescue => e
    failure(e)
    raise e
  end

  def success
    LogicLog.create!(
        source: "#{self.class} #{provider_job_id}",
        level: 0,
        msg: 'Success'
    )
  end

  def failure(error)
    LogicLog.create!(
        source: "#{self.class} #{provider_job_id}",
        level: 0,
        msg: "Error: #{error.message}\nchanges: #{arguments[2]}\nclass: #{arguments[0]}\nqueue: #{queue_name}\nsql: #{arguments[1]}"
    )
    raise error
  end

  def set_audit_log_data
    ::PaperTrail.request.whodunnit = who_is[:whodunnit]
    ::PaperTrail.request.controller_info = who_is[:controller_info]
  end
end
```

Comment (senid231) [reply lib/batch_jobs_log.rb:36]: and add `config/initializers/delayed_job.rb` to make it work

```
class Delayed::ActiveJobPlugin < Delayed::Plugin
  callbacks do |lifecycle|
    lifecycle.before(:invoke_job) do |job|
      job.payload_object.job_data['provider_job_id'] = job.id
    end
  end
end

Delayed::Worker.plugins << Delayed::ActiveJobPlugin
```

**Comment by senid231** [comment]
https://github.com/collectiveidea/delayed_job/issues/1118

File: spec/jobs/async_batch_destroy_job_spec.rb [hunk 11]
```diff
@@ -44,6 +44,31 @@
...
+
+        context 'should write record about' do
+          let(:sql_query) { Contractor.where(id: contractor_alone.id).to_sql }
+          it 'success performed job' do
+            expect { subject }.to change(LogicLog, :count).by 1
+            expect(LogicLog.last.msg).to start_with 'Success'
+          end
+        end
+
+        context do
```
Comment (gigorok) [inline spec/jobs/async_batch_destroy_job_spec.rb:62]:
```suggestion
        context 'when the job raise an error' do
```

File: spec/jobs/async_batch_update_job_spec.rb [hunk 12]
```diff
@@ -69,6 +69,28 @@
...
+
+        context 'should write record about' do
+          let(:changes) { { next_interval: 4 } }
+          it 'success performed job' do
+            expect { subject }.to change(LogicLog, :count).by 1
+            expect(LogicLog.last.msg).to start_with 'Success'
+          end
+        end
+
+        context do
```
Comment (gigorok) [inline spec/jobs/async_batch_update_job_spec.rb:84]:
```suggestion
        context 'when the job raise an error' do
```

File: config/initializers/delayed_job.rb [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+
+class Delayed::ActiveJobPlugin < Delayed::Plugin
+  callbacks do |lifecycle|
+    lifecycle.before(:invoke_job) do |job|
+      job.payload_object.job_data['provider_job_id'] = job.id
+    end
+  end
+end
+
+Delayed::Worker.plugins << Delayed::ActiveJobPlugin
```
Comment (gigorok) [inline config/initializers/delayed_job.rb:11]: please move to this initializer this config
Delayed::Worker.destroy_failed_jobs = false
from config/initializers/yeti.rb

**Pull request merged**